[docs] Rewrite Supabase guide around eas integrations:supabase:connect (#48558)
expo/expo: docs/pages/guides/using-supabase.mdx

---
title: Using Supabase
description: Add a Postgres database and user authentication to your React Native app with Supabase.
platforms: ['android', 'ios']
---

import { BoxLink } from '~/ui/components/BoxLink';
import { Collapsible } from '~/ui/components/Collapsible';
import { Prerequisites, Requirement } from '~/ui/components/Prerequisites';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';

> [Supabase](https://supabase.com/?utm_source=expo&utm_medium=referral&utm_term=expo-react-native) is a Backend-as-a-Service app development platform built on Postgres. It [generates a REST API](https://supabase.com/docs/guides/api?utm_source=expo&utm_medium=referral&utm_term=expo-react-native) from your database and uses [row level security](https://supabase.com/docs/guides/auth/row-level-security?utm_source=expo&utm_medium=referral&utm_term=expo-react-native) to protect the data, so your React Native app can query that API directly, with no server in between.

Reviewer: ```suggestion
[Supabase](https://supabase.com/?utm_source=expo&utm_medium=referral&utm_term=expo-react-native) is a Backend-as-a-Service (BaaS) app development platform built on Postgres. It [generates a REST API](https://supabase.com/docs/guides/api?utm_source=expo&utm_medium=referral&utm_term=expo-react-native) from your database and uses [row level security](https://supabase.com/docs/guides/auth/row-level-security?utm_source=expo&utm_medium=referral&utm_term=expo-react-native) to protect the data, so your React Native app can query that API directly, with no server in between.
```

It's a general term and can be abbreviated. We should keep it.

The EAS CLI integration automates the standard setup: authorizing your Supabase account, creating or linking a project, installing the SDK, and writing your environment variables. You can also [set it up manually](#manual-setup) and use the rest of this guide unchanged.

<Prerequisites>
  <Requirement title="Expo account">
    Sign up for an [Expo account](https://expo.dev/signup).
  </Requirement>
  <Requirement title="EAS CLI">Install EAS CLI globally with `npm install -g eas-cli`.</Requirement>
  <Requirement title="Expo project linked to EAS">
    Create an Expo project and link it to EAS with `eas init`.
  </Requirement>
  <Requirement title="Supabase account">
    Sign up for a [Supabase
    account](https://supabase.com/dashboard/sign-up?utm_source=expo&utm_medium=referral&utm_term=expo-react-native).
  </Requirement>
</Prerequisites>

## What you'll learn

- [Install and configure Supabase](#install-and-configure-supabase) in your React Native app
- [Add authentication](#add-authentication) with the client you create in Step 2
- [Set up environments](#set-up-environments) for local development, Production, and Preview

> - [Manage the integration](#manage-the-integration) and [troubleshooting](#troubleshooting)

Reviewer: 💡 **nit** `links-and-headings.md#titles-and-headings`

"Manage the integration and troubleshooting" pairs a verb with a noun; consider "[Manage the integration](#manage-the-integration) and [troubleshoot common problems](#troubleshooting)".

```suggestion
- [Manage the integration](#manage-the-integration) and [troubleshoot common problems](#troubleshooting)
```

<!-- docs-review -->

## Install and configure Supabase

<Step label="1">

### Run the `connect` command

Run the following command in your project directory:

<Terminal cmd={['$ eas integrations:supabase:connect']} />

With no project linked, the command creates a new Supabase project. To use a Supabase project you already have, link it instead:

<Terminal cmd={['$ eas integrations:supabase:connect --link <project-ref-or-url>']} />

This command:

- Opens your browser to authorize Supabase, then continues after you approve.
- Asks which Supabase **organization** to use, if your account has more than one.

> - Asks for a **region** from Americas, Europe / Middle East / Africa, and Asia Pacific, then creates the project and waits until it's ready. The region sets your data residency and can't be changed after the project is created.

Reviewer: ✏️ **style** `mechanics.md#no-space-around-`

No spaces around the slash: "Europe / Middle East / Africa" should be "Europe/Middle East/Africa".

```suggestion
- Asks for a **region** from Americas, Europe/Middle East/Africa, and Asia Pacific, then creates the project and waits until it's ready. The region sets your data residency and can't be changed after the project is created.
```

<!-- docs-review -->

</Step>

<Step label="3">

### Create a table

Run `eas integrations:supabase:dashboard` to open your linked project, or open it from the Supabase dashboard. Select **SQL Editor**, then run:

```sql Supabase SQL Editor
create table public.todos (
  id bigint generated always as identity primary key,
  title text not null
);
alter table public.todos enable row level security;
create policy "Anyone can read todos" on public.todos for select using (true);
grant select on public.todos to anon, authenticated;
insert into public.todos (title) values ('Hello from Supabase');
```

Requests use the `anon` role when signed out and `authenticated` after sign-in, and the policy decides which rows those roles can read. Without a policy, a select returns an empty array and no error. To let the app write, add an insert policy.

> The grant is a safeguard rather than a requirement. Hosted projects already grant `select`, `insert`, `update`, and `delete` on new tables in `public` to both roles, but [Supabase is making those grants opt-in](https://supabase.com/docs/guides/database/hardening-data-api?utm_source=expo&utm_medium=referral&utm_term=expo-react-native), and a project with them revoked fails with `permission denied for table todos`. Granting a privilege the table already has changes nothing.

Reviewer: ✏️ **style** `prose.md#sentence-construction`

The second sentence chains three clauses: "...to both roles, but Supabase is making those grants opt-in, and a project with them revoked fails...". Prefer separate sentences, for example split before "but" and continue with "However,".

```suggestion
The grant is a safeguard rather than a requirement. Hosted projects already grant `select`, `insert`, `update`, and `delete` on new tables in `public` to both roles. However, [Supabase is making those grants opt-in](https://supabase.com/docs/guides/database/hardening-data-api?utm_source=expo&utm_medium=referral&utm_term=expo-react-native), and a project with them revoked fails with `permission denied for table todos`. Granting a privilege the table already has changes nothing.
```

<!-- docs-review -->

</Step>

<Step label="4">

### Verify the configuration

Replace the contents of your first screen with a query against the table:

```tsx src/app/index.tsx
import { useEffect, useState } from 'react';
import { Text, View } from 'react-native';
import { supabase } from '@/lib/supabase';

export default function Index() {
  const [titles, setTitles] = useState<string[]>([]);

  useEffect(() => {
    supabase
      .from('todos')
      .select()
      .then(({ data, error }) => {
        if (error) {
          setTitles([error.message]);
          return;
        }
        setTitles(data.map(todo => todo.title));
      });
  }, []);

  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      {titles.map(title => (
        <Text key={title}>{title}</Text>
      ))}
    </View>
  );
}
```

Start your app:

<Terminal cmd={['$ npx expo start']} />

If the screen shows `Hello from Supabase`, your client works. An empty screen means no policy allows the read. An error message means something else, so see [Troubleshooting](#troubleshooting).

> Everything in this guide runs in Expo Go. You need a [development build](/develop/development-builds/introduction/) once you pass options to the `expo-sqlite` plugin or add other native libraries.

Reviewer: ✏️ **style** `terminology.md#expo-product-and-feature-names`

For Expo Go, the docs say "test in Expo Go" rather than "runs in Expo Go": try "You can test everything in this guide in Expo Go."

```suggestion
You can test everything in this guide in Expo Go. You need a [development build](/develop/development-builds/introduction/) once you pass options to the `expo-sqlite` plugin or add other native libraries.
```

<!-- docs-review -->

</Step>

For the concepts behind these steps, see the Supabase database overview:

<BoxLink
  title="Supabase database overview"
  description="Tables, row level security policies, and realtime updates."
  href="https://supabase.com/docs/guides/database/overview?utm_source=expo&utm_medium=referral&utm_term=expo-react-native"
/>

## Add authentication

> The client from [Step 2](#create-the-supabase-client) already stores sessions, so email and password sign-in needs no extra configuration. New Supabase projects confirm email addresses by default, so your first `signUp` returns `data.user` with `data.session` set to `null` until the address is confirmed or you turn off **Confirm email** in your project's email provider settings.

Reviewer: 💡 **nit** `prose.md#sentence-construction`

Consider splitting the second sentence at ", so": "New Supabase projects confirm email addresses by default. Your first `signUp` returns..."

```suggestion
The client from [Step 2](#create-the-supabase-client) already stores sessions, so email and password sign-in needs no extra configuration. New Supabase projects confirm email addresses by default. Your first `signUp` returns `data.user` with `data.session` set to `null` until the address is confirmed or you turn off **Confirm email** in your project's email provider settings.
```

<!-- docs-review -->

</Collapsible>

<Collapsible summary="Project reference ID doesn't work">

The project must belong to the Supabase organization you connected. See [Finding your project reference ID](#finding-your-project-reference-id) for the accepted formats.

</Collapsible>

<Collapsible summary="Environment variables don't update">

After `connect` writes them, reload the app so it picks up the new values. If the app still reads the old value, stop the development server and run `npx expo start` again.

</Collapsible>

<Collapsible summary="Table doesn't exist right after you create it">

Supabase caches your database schema, so a query right after you create a table can fail with `Could not find the table 'public.todos' in the schema cache`. Re-run it; the cache refreshes on its own.

</Collapsible>

<Collapsible summary="Permission denied for a table">

Add the grant your query needs for both roles, such as `grant select on public.todos to anon, authenticated`. A policy alone isn't enough. Unlike a missing policy, which returns an empty array, this fails with `error.code` `42501`.

</Collapsible>

<Collapsible summary="Supabase authorization stopped working">

> If you revoked Expo's access in Supabase, run `eas integrations:supabase:connect --reauth`. This clears the stored connection and project link, reopens the browser, then asks whether to link or provision, so have your reference ID ready. Your Supabase projects aren't affected. `--reauth` needs a browser, so it fails in non-interactive mode.

Reviewer: 💡 **nit** `prose.md#core-principles`

"provision" is the one place the guide doesn't say "create"; consider "asks whether to link a project or create one".

```suggestion
If you revoked Expo's access in Supabase, run `eas integrations:supabase:connect --reauth`. This clears the stored connection and project link, reopens the browser, then asks whether to link a project or create one, so have your reference ID ready. Your Supabase projects aren't affected. `--reauth` needs a browser, so it fails in non-interactive mode.
```

<!-- docs-review -->

</Collapsible>

<Collapsible summary="Extra project created without environment variables">

If `connect --environment` creates a project and then fails to write the environment variables, the command prints the project URL and publishable key. Save those values with [`eas env:set`](/eas/environment-variables/).

> **warning** Don't re-run `connect --environment` in this case. It creates another project, which counts against your plan limit.

</Collapsible>

> ## Learn more

Reviewer: ```suggestion
## Further reading
```

Let's change this so that Google Search Console and Core Web Vitals don't complain about it.
